invisible_cities/reco/xy_algorithms_c.pyx

```diff
@@ -0,0 +1,162 @@
+                    lm_radius can always be set to 0 mm, but setting it to 15 mm (slightly larger
+                than 10mm * sqrt(2)), should not hurt.
+
+    new_lm_radius = radius, find a new cluster by calling barycenter() on pos/qs of SiPMs within
+                    new_lm_radius of new_local_maximum
+    returns
+    c    : a list of Clusters
+
+    Usage Example
+    In order to create each Cluster from a 3x3 block of SiPMs (where the center SiPM has more
+    charge than the others), one would call:
+    corona(pos, qs,
+           Qthr           =  K1 * units.pes,
+           Qlm            =  K2 * units.pes,
+           lm_radius      =  0  * units.mm , # must be 0
+           new_lm_radius  =  15 * units.mm , # must be 10mm*sqrt(2) or some number slightly larger
+           msipm          =  K3)
+    """
+
+    if len(pos) == 0: raise SipmEmptyList
+    if  sum(qs) == 0: raise SipmZeroCharge
+
+    cdef np.ndarray[long, ndim=1] above_threshold
+    above_threshold = np.where(qs >= Qthr)[0]            # Find SiPMs with qs at least Qthr
+    pos, qs = pos[above_threshold], qs[above_threshold]  # Discard SiPMs with qs less than Qthr
+
+    if len(pos) == 0: raise SipmEmptyListAboveQthr
+    if  sum(qs) == 0: raise SipmZeroChargeAboveQthr
+
+    # if lm_radius or new_lm_radius is negative, just call overall barycenter
+    if lm_radius < 0 or new_lm_radius < 0:
+        return barycenter(pos, qs)
+
+    cdef list c = []
+    cdef np.ndarray[long, ndim=1] within_lm_radius
+    cdef np.ndarray[long, ndim=1] within_new_lm_radius
+    cdef tuple new_local_maximum
+    cdef int hottest_sipm
+    # While there are more local maxima
+    while len(qs) > 0:
+
+        hottest_sipm = np.argmax(qs)       # SiPM with largest Q
+        if qs[hottest_sipm] < Qlm: break   # largest Q remaining is negligible
+
+        # find new local maximum of charge considering all SiPMs within lm_radius of hottest_sipm
+        within_lm_radius   = get_nearby_sipm_inds(tuple(pos[hottest_sipm]), lm_radius, pos, qs)
+        new_local_maximum  = barycenter(pos[within_lm_radius], qs[within_lm_radius])[0].XY
+
+        # find the SiPMs within new_lm_radius of the new local maximum of charge
+        within_new_lm_radius = get_nearby_sipm_inds(new_local_maximum, new_lm_radius, pos, qs)
+
+        # if there are at least msipms within_new_lm_radius, get the barycenter
+        if len(within_new_lm_radius) >= msipm:
+            c.extend(barycenter(pos[within_new_lm_radius], qs[within_new_lm_radius]))
+
+        # delete the SiPMs contributing to this cluster
+        pos, qs = discard_sipms(within_new_lm_radius, pos, qs)
+
+    if not len(c): raise ClusterEmptyList
+
+    return c
+
+
+cpdef discard_sipms(np.ndarray[  long, ndim=1] sis,
+                    np.ndarray[double, ndim=2] pos,
+                    np.ndarray[double, ndim=1] qs):
+    return np.delete(pos, sis, axis=0), np.delete(qs, sis)
+
+
+cpdef get_nearby_sipm_inds(tuple cs, float d,
```

Comment:
why do we use `cs` as tuple here? I see that in the first call to this function we explicitly cast some value to tuple.